### ISSUE TITLE
[Bug] No retry on Piston server

### ISSUE THREAD

**Issue by tr7zw** [Kir-Antipov/mc-publish#131]
### Description

When Mojangs metadata server is having issues, the entire CI might randomly fail because there is no retry logic.

### Expected Behavior

Re-attempt a few times to load the json, same as when the Modrinth/CF upload fails due to API/S3 errors.

### Actual Behavior

> Run Kir-Antipov/mc-publish@v3.3.0
Error: FetchError: request to https://piston-meta.mojang.com/mc/game/version_manifest_v2.json failed, reason: 

(That is the entire log, there is no reason provided)

### Version

3.3.0

### Environment

GitHub Actions

**Comment by lattaai2** [comment]
Hello, I tried to fix the issue

This is what I did:
I have implemented a retry mechanism for the MojangApiClient to handle potential issues with Mojang's metadata server. This solution adds retry logic to the `getAllMinecraftVersions` method, which is the core method for fetching Minecraft version data.

Here's a summary of the changes:

1. I've imported the `retry` function from `@/utils/async-utils`.
2. I've added a new private method `fetchVersionManifest` that encapsulates the fetching logic with retry mechanism.
3. I've updated the `getAllMinecraftVersions` method to use the new `fetchVersionManifest` method.
4. The retry logic will attempt to fetch the version manifest up to 3 times, with a 1-second delay between attempts.
5. I've also added error logging using the default logger.

These changes should make the MojangApiClient more resilient to temporary server issues while maintaining its current functionality.

I also created pull request: https://github.com/Kir-Antipov/mc-publish/pull/132


> [!CAUTION]  
> Disclaimer: This fix was created by Latta AI and you should never merge before you check the correctness of generated code!
> 
> The fix provided by Latta AI might not be complete and it can serve as an inspiration.

---
This bug was fixed for free by Latta AI - https://latta.ai/ourmission

If you no longer want Latta AI to attempt fixing issues on your repository, you can block this account.

**Comment by tr7zw** [comment]
First of all, that's a first with this AI pr (also lmao about ``If you no longer want Latta AI to attempt fixing issues on your repository, you can block this account.``). But it actually seems correct from a glance.

For a bit of context, first had the EntityCulling deploy crashing halfway through:
https://github.com/tr7zw/EntityCulling/actions/runs/11635994760/job/32406501958
so I manually uploaded the missing 15 or so files to Modrinth/Curseforge, but then at the same time 3d skin layers failed and only got 9 out of the 62? files published ( https://github.com/tr7zw/3d-Skin-Layers/actions/runs/11636424329/job/32407800763 ), and I really don't feel like publishing 53 files manually 😵‍💫.

**Comment by Kira-NT** [comment]
That's a great idea!

Honestly, I hadn't thought of it before because **usually** the Mojang API has a pretty good uptime. The chances of running into issues with it seemed about as likely as a GitHub Runner losing internet connectivity during the publishing process *(which, ironically, happened to me just a few days ago :D)*

This is actually the first outage of this kind I've ever seen.

> First of all, that's a first with this AI pr

*Sigh*. Just ignore this garbage *(or, better yet, report it)*

> and I really don't feel like publishing 53 files manually

Yikes, that's unfortunate. At this point, it might be easier to delete the versions that have already been published and try again a little later, once Mojang gets their stuff together.

**Comment by tr7zw** [comment]
Right now the plan is to just make it version 1.7.0.1 and retry to publish it tomorrow. Deleting releases is a bit tricky, and I can't re-run the publish since mcpublish re-trying to publish these releases will cause issues too. (Actually, if it could just skip the publish if a file with 1:1 the same name/target version/version/mod loader exists, that would be really helpful to recover from these half-baked releases due to network related crashes)

**Comment by Kira-NT** [comment]
> Actually, if it could just skip the publish if a file with 1:1 the same name/target version/version/mod loader exists, that would be really helpful to recover from these half-baked releases due to network related crashes

I really do need to implement something like this to make recovery from these unfortunate situations easier. It certainly should be possible with Modrinth *(god bless the Modrinth development team and their actually competent API implementation)*, though I do have some doubts about the CurseForge Upload API in that regard.

**Comment by tr7zw** [comment]
Yea it's tricky. Makes me with Github Actions had the same way of working as Jenkins. On a Jenkins I could just re-trigger the job with a modified build file(from the webinterface) where I comment out the tasks that are already done, so it just finishes the broken ones.

**Comment by Kira-NT** [comment]
Yup, Modrinth has [an endpoint](https://docs.modrinth.com/api/operations/versionfromhash/) to check if a file has already been uploaded to their servers, making it super simple to skip duplicate uploads in the event of re-runs.

---

CurseForge, on the other hand, doesn't offer a direct route for this, but it is possible to retrieve [**all** the mod's files](https://docs.curseforge.com/rest-api/#get-mod-files) and then check their hashes:

```json
"hashes": [
  {
    "value": "a0193dd17bf8a98868cd53edeffc144962770c95",
    "algo": 1
  },
  {
    "value": "42020dbc2c86e93f7f8e2d28a349c4a4",
    "algo": 2
  }
]
```

It seeems that `"algo": 1` represents SHA-1, which is also used by Modrinth, so at least that's good.

**Comment by Kira-NT** [comment]
> On a Jenkins I could just re-trigger the job with a modified build file(from the webinterface) where I comment out the tasks that are already done, so it just finishes the broken ones.

Like for real! It would be a lifesaver in so many situations. GitHub already has an option to rerun only failed jobs, but nothing for rerunning just the failed steps *(and the steps they depend on)*. They could even pair this with some fancy-pants interface that lets us select the specific steps that need to be rerun, so we wouldn't have to manually edit the workflow file to get it done. But oh well, maybe in another 10 years they'll sort that out :D

**Comment by tr7zw** [comment]
Maybe related, had piston fail last night again, thankfully on the first one. Now retried, and it failed halfway here:
![grafik](https://github.com/user-attachments/assets/12772e04-224a-4093-94ef-9d9de7400e72)
Shouldn't that have a retry as well?

**Comment by Kira-NT** [comment]
Damn, these server unavailability problems have certainly become much more common than they were back in the ol' days. I guess I need to create a retry middleware for my `fetch` implementation and set it up for basically every API client.

**Comment by tr7zw** [comment]
That would be amazing, bestcase before the 1.21.4 release wave starts 😅.

**Comment by Kira-NT** [comment]
I don't think I'm going to make it, to be quite frankly honest with you! The issue is that I've been transitioning mc-publish away from its current distribution model *(due to GitHub's unhinged handling of Node runners)* and moving away from Node.js in general for quite some time now, while simultaneously cleaning up a backlog of bug reports and feature requests. As a result, my local codebase now differs significantly from the current upstream state of the project, which makes it pretty difficult to rebase and backport features to `v3.x` in the meantime.

So, yeah, with that in mind, I don't think I'll be able to finish everything before 1.21.4 drops. On the bright side, though, 1.22 should *(hopefully)* be a smooth sailing after all that! :D

**Comment by tr7zw** [comment]
Ah well, then let's hope everything goes smoothly 🫡

Edit:
![grafik](https://github.com/user-attachments/assets/31b98852-8fdd-4550-8f22-588830453b8f)
*sigh* first one
